nep-12.mediawiki

```diff
@@ -0,0 +1,2130 @@
+  name: string;
+  version: string;
+  author: string;
+  email: string;
+  description: string;
+  needsStorage?: boolean;
+  dynamicInvoke?: boolean;
+  isPayable?: boolean;
+  parameterList: string;
+  returnType: string;
+  code: string;
+  network?: string - Network to submit this request to. If omitted, will default to network the wallet is currently set to.
+  broadcastOverride?: boolean; // In the case that the dApp would like to be responsible for broadcasting the signed transaction rather than the wallet provider
+}
+</pre>
+
+
+'''Success return value'''
+
+<pre>
+interface DeployOutput {
+  txid: string;
+  nodeUrl?: string; // The node which the transaction was broadcast to. Returned if transaction is broadcast by wallet provider
+  signedTx?: string; // The serialized signed transaction. Only returned if the broadcastOverride input argument was set to True
+}
+</pre>
+
+
+'''Error return value'''
+
+<pre>
+interface Error {
+  type: string;
+  description: string;
+  data: string;
+}
+</pre>
+
+
+'''Example'''
+
+<pre>
+deploy({
+  name: 'Hello world!',
+  version: 'v0.0.1',
+  author: 'John Smith',
+  email: 'info@o3.network',
+  description: 'My first contract.',
+  needsStorage: true,
+  dynamicInvoke: false,
+  isPayable: false,
+  parameterList: '0710',
+  returnType: '05',
+  code: '53c56b0d57616b652075702c204e454f21680f4e656f2e52756e74696d652e4c6f6761006c7566',
+})
+.then(({txid, nodeUrl}: InvokeOutput) => {
+  console.log('Deploy transaction success!');
+  console.log('Transaction ID: ' + txid);
+  console.log('RPC node URL: ' + nodeUrl);
+})
+.catch(({type: string, description: string, data: any}) => {
+  switch(type) {
+    case UNKNOWN_ERROR:
+      console.log(description);
+      break;
+  }
+});
+</pre>
+
+
+'''Provider Request Handling'''
+
+Upon receiving this request
+* Run a testinvoke on the deploy transaction to get the gas cost in order to deploy the contract
+* Display the deploy input parameters to the user along with the associated deployment fee
+* Sign transaction upon receiving confirmation from user
+* If the "broadcastOverride" input argument is set to True, return the serialized signed transaction to the dApp
+* Else broadcast the RPC request
+* Return the transaction Id and the url of the rpc node which the tx was broadcast to
+
+Example
+<pre>
+{
+  txid: 'ed54fb38dff371be6e3f96e4880405758c07fe6dd1295eb136fe15f311e9ff77',
+  nodeUrl: 'http://seed7.ngd.network:10332',
+}
+</pre>
+
+
+===Events===
+
+dApps can listen for events emitted by the wallet provider using the `addEventListener` method.
+
+====addEventListener====
+
+'''Method Interface'''
+
+<pre>
+function addEventListener(event: Event, callback: Function): Void
+</pre>
+
+'''Input arguments'''
+
+<pre>
+enum Event {
+  READY = 'READY',
+  ACCOUNT_CHANGED = 'ACCOUNT_CHANGED',
+  NETWORK_CHANGED = 'NETWORK_CHANGED',
+  CONNECTED = 'CONNECTED',
+  DISCONNECTED = 'DISCONNECTED',
+  BLOCK_HEIGHT_CHANGED = 'BLOCK_HEIGHT_CHANGED',
```

Comment:
BLOCK_HEIGHT_CHANGED, or BLOCKCHAIN_HEIGHT_CHANGED (or just HEIGHT_CHANGED) ?

Reply:
I have no opposition to this. As long as all other wallet providers and dapps are okay with making the update.

Reply:
Ok.